docs/proposals/20220228-advanced-cpuset-manger.md

```diff
@@ -0,0 +1,94 @@
+authors:
+  - "@szy441687879"
+reviewers:
+  - "@yan234280533"
+  - "@mfanjie"
+creation-date: 2022-02-28
+last-updated: 2022-02-28
+status: provisional
+---
+
+# Advanced CPUSet Manager
+
+- Provide a new mechanism to manage cpuset bypass, to do extra things that kubelet cannot do, such as revise cpuset when pod running, share the cpuset assigned with others and relax restrictions of cpuset allocation.
+
+## Table of Contents
+
+<!-- TOC -->
+
+- [Advanced CPUSet Manager](#advanced-cpuset-manager)
+    - [Table of Contents](#table-of-contents)
+    - [Motivation](#motivation)
+        - [Goals](#goals)
+        - [Non-Goals/Future Work](#non-goalsfuture-work)
+    - [Proposal](#proposal)
+        - [Relax restrictions of cpuset allocation](#relax-restrictions-of-cpuset-allocation)
+        - [Add new annotation to describe the  requirement of cpuset contorl manger](#add-new-annotation-to-describe-the--requirement-of-cpuset-contorl-manger)
+        - [Advanced CPU Manager component](#advanced-cpu-manager-component)
+        - [User Stories](#user-stories)
+            - [Story 1](#story-1)
+            - [Story 2](#story-2)
+        - [Risks and Mitigations](#risks-and-mitigations)
+
+<!-- /TOC -->
+## Motivation
+Some latency-sensitive applications have lower lantency and cpu usage when running with specific cores, which results in fewer context switchs and higer cache affinity.
+But kubelet will always exclude assigned cores in shared cores, which may waste resources.Offline and other online pods can running on the codes actually. In our experiment, for the most part, it is barely noticeable for performance of service.
+
+### Goals
+
+- Provide a new mechanism to manage cpuset bypass
+- Provide a new cpuset control way "shared"
+- Allow revise cpuset when pod running
+- Relax restrictions of cpuset allocation
+
+
+### Non-Goals/Future Work
+
+- Solve the conflicts with kubelet static cpuset control, you need to set kubelet cpuset manager to "none"
+- Numa manager will support in future, CCX/CCD manager can also be considered
+
+## Proposal
+### Relax restrictions of cpuset allocation
+Kubelet allocate cpus for containers included in a pod of "Guaranteed" QoS class and the container's resource limit for the CPU resource is an integer greater than or equal to one.
+
+With Crane, containers only to neeed CPU resource limit is an integer greater than or equal to one and equal to cpu resource request.
+
+### Add new annotation to describe the  requirement of cpuset contorl manger
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  annotations:
+    qos.gocrane.io/cpu-manager: none/exclusive/share
+```
+Provide three polices for cpuset control:
+- none: containers of this pod shares a set of cpus not allocated top exclusive policy containers
+- exclusive:  containers of this pod exclude the CPUs allocated, other containers not allowed to use.
+- share: containers of this pod runs in the CPUs allocated, but other containers can alse use.
```

Comment:
typo.

Reply:
Kubelet allocate cpus for containers should meet the conditions:

1. requests and limits are specified for all the containers and they are equal

2. the container's resource limit for the limit of CPU is an integer greater than or equal to one and equal to request request of CPU.

In Crane, only need to meet condition No.2

Reply:
done